tiiuae/ghaf
# Add login manager, magnification, reuse.toml, and desktop improvements

- Adds a login manager and a graphical greeter in Ghaf, with option for autologin (by default for now). This removes redundant services such as ghaf-session, and seatd (as replaced with an official NixOS service option).
- Adds magnification in labwc, which can be used for accessibility, development, and demo purposes.
- Switch to reuse.toml. Reuse specification version 3.2 deprecates support of dep5. Also more explicitly define the licenses for patches.
- Improves launcher speed and adds a keybind (super/windows left key). Set time to 24h.
- Improve audio keybindings latency, and desktop theme.
- Run 'autostart' application as systemd user services.

## Description of changes

- Adds greetd login manager
  - This runs as a separate user 'greeter' which is in 'video' group.
  - Use gtkgreet as a greeter, customised to match our style guide.
  - We may switch to something more customisable, such as ReGreet.
  - Uses labwc compositor to run the greeter
  - Brightness controls, alternative keyboard layouts works on the login screen.
  - Not possible to launch Terminal from greeter user.
  - labwc supports multi-monitor well, as well as `wlr-layer-shell-unstable` for gtkgreet.
  - Replace custom seatd service with `services.seatd` as it is now available in NixOS.
  - Set wlroots environment variables using `environment.sessionVariables`.
  - This allows other compositor to work properly, but this PR doesn't add any new compositors.
  - Fix renderer option for Orin (previously it seemed to fallback correctly even with typo).
- Add keybindings for magnification.
  - Super+z for toggling magnification.
  - Super+- for zoom out.
  - Super++ (without pressing Shift) for zoom in.
- Switch to reuse.toml
  - Use reuse.toml instead of dep5 (Debian-style), as it has been deprecated.
  - Keep all license information in `REUSE.toml` so it is easier to define in a single place.
  - Include licenses for the patches included in this repository.
  - License assets for docs under CC-BY-SA-4.0 instead of Apache-2.0.
- Drawer improvements
  - Use nwg-drawer memory resident feature, so it stays in background after login. This improves launcher speed.
  - Adds keybind (windows/Super Left) key for opening drawer.
  - Sets clock to 24h, other cleanup.
- Improve desktop
  - Only right click should show debug menu, on debug builds.
  - Added drop shadows for windows.
  - Mouse should use regular scrolling instead of natural. Touchpad uses natural scrolling.
  - Removal of 'Ghaf Platform' desktop menu element (used to do nothing).
  - Cleanup of theme definitions, and improve theme for snapping.
- Run autostart applications as systemd user services.
  - This allows us to view the logs of the user services in journald.
  - It is easier to monitor these services, and restart them.
  - Services include: waybar, swaybg, mako, ghaf-launcher (nwg-drawer), and autolocking.
  - These services properly quit when the user logs out.

## Checklist for things done






## Testing



- By default, you will be logged in automatically.
- You can log out by clicking Log Out in app drawer, which should bring you to the login screen.
- Setting `ghaf.graphics.labwc.autologinUser = null;` would disable autologin, and show the login screen on boot.
- Login, logout multiple times. Should not cause system to go in bad state.
- In login screen, plug and unplug external monitors. Should not glitch or crash.
  - Login prompt is shown on monitor with cursor, moving cursor to other monitor would move the prompt.
  - Other monitors should only show clock on the centre.
- All applications should be launchable after re-logins.
- Boot time shouldn't be significantly affected.
- Fingerprint authentication should work with both lock screen and login screen. Login screen should not hog access to fingerprint reader after login.
  - With same previous caveats with gtklock.
- Test zoom keybindings on desktop, and login screen (it should work there too).
- Run `reuse lint`, should pass with no issues.
- Open/close launcher multiple times using taskbar, should be fast and responsive.
- Open/close launcher using Super Left/Windows key, should be fast.
- The drawer should not impede on the functionality of zoom (since it is using the same modifier key).
